### PR TITLE
Msalem/add scheduler function

### PR DESCRIPTION
**Pull request by msalemcode**
Adding new feature to schedule metered Emitting events. This PR will contain change to every application layer to reflect the newly added entity. Here is summary for changes

- SQL Server : Added 2 new tables + 1 new view for scheduler
- DAL : Added new entity to context and added contract and repository
- Business Service: Added Service and Model
- Publisher site: New View and controller for scheduler 
- New Function for scheduler

### CODE REVIEW

**Comment by code4clouds** [comment]
@msalemcode  Would you attach screenshot of this features?

**Comment by tiffwchen** [comment]
For ./docs/Installation-Instructions.md, please change the title from "Installating" to "Installing"

File: src/SaaS.SDK.Client.DataAccess/Services/MeteredDimensionsRepository.cs [hunk 1]
```diff
@@ -51,6 +51,7 @@ public IEnumerable<MeteredDimensions> Get()
         /// <returns> Metered Dimensions.</returns>
         public MeteredDimensions Get(int id)
         {
+
```
Comment (tiffwchen) [inline src/SaaS.SDK.Client.DataAccess/Services/MeteredDimensionsRepository.cs:54]: Remove this newline

File: src/SaaS.SDK.Client.DataAccess/Services/MeteredPlanSchedulerManagementRepository.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// The disposed.
+        /// </summary>
+        private bool disposed = false;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MeteredPlanSchedulerManagementRepository"/> class.
+        /// </summary>
+        /// <param name="context">The this.context.</param>
```
Comment (tiffwchen) [inline src/SaaS.SDK.Client.DataAccess/Services/MeteredPlanSchedulerManagementRepository.cs:29]: "The SaasKitContext" instead?

**Review by tiffwchen — CHANGES_REQUESTED** [review]

File: src/SaaS.SDK.PublisherSolution/Views/Home/Index.cshtml [hunk 3]
```diff
@@ -109,6 +109,38 @@
...
             </div>
         </li>
+        @if (this.TempData["SupportMeteredBilling"]!=null)
+        {
+            <li id="onedash-menu-homepage5">
+                <a class="programCardLink text-capitalize" asp-area="" asp-controller="Scheduler" asp-action="Index">Scheduler</a>
+            <div class="programCard">
+                <div class="programCardContentWrapper">
+                    <div class="programCardTitle minline">
+                        <div class="programCardIcon5 wrap-win-icon-RecruitmentManagement">
```
Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Views/Home/Index.cshtml:119]: I have a concern that scheduler and offers icons are similar which is not good for usability: 
image
 Would it be possible to replace it with another bootstrap icon, for example such as 
image

File: src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                <span class="cm-section-heading">Add New Scheduled Trigger for Metered Plan</span>
+            </div>
+        </div>
+
+
+        <form method="post" id="frmSchedulerEvents" asp-action="AddNewScheduledTrigger" asp-controller="Scheduler">
+            <div style="align-content:center">
+                <table>
+                    <tr>
+                        <th>Scheduler Name&nbsp;&nbsp;</th>
```
Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:44]: can you add check for empty for schedule name?

Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:44]: can you add check for empty for schedule name?

File: src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    <tr>
+                        <th>Subscription</th>
+                        <th>@Html.DropDownListFor(s => s.SelectedSubscription, Model.SubscriptionList, "-- Select Subscription --", new { @class = "form-control", required = "required"})</th>
+                    </tr>
+                    <tr>
+                        <th>Dimenstion</th>
+                        <th>@Html.DropDownListFor(s => s.SelectedDimension, Model.DimensionsList, "-- Select Dimension --", new { @class = "form-control", required = "required" })</th>
+                    </tr>
+                    <tr>
+                        <th>Quantity</th>
```
Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:56]: can you add check for 0 for quantity?

Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:56]: can you add check for 0 for quantity?

File: src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                var requestJson = JsonSerializer.Serialize(subscriptionUsageRequest);
+                var responseJson = string.Empty;
+                try
+                {
+                    Console.WriteLine($"Execute EmitUsageEventAsync for request {requestJson}");
+                    meteringUsageResult = billingApiService.EmitUsageEventAsync(subscriptionUsageRequest).ConfigureAwait(false).GetAwaiter().GetResult();
+                    responseJson = JsonSerializer.Serialize(meteringUsageResult);
+                    Console.WriteLine($"Got the following result{responseJson}");
+                }
+                catch (MarketplaceException mex)
```
[the remark this answers is not in the document]
Comment (msalemcode) [reply src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs:102]: Not sure what do you mean here However this code is copy from existing code for Client SDK

[the remark this answers is not in the document]
Comment (msalemcode) [reply src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs:102]: Added appropriate exception name

File: src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml (hunk 5 above)
Comment (msalemcode) [reply src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:56]: Add number input type

**Review by tiffwchen — APPROVED** [review]

File: src/SaaS.SDK.MeteredTriggerJob/SaaS.SDK.MeteredTriggerJob.csproj [hunk 7]
```diff
@@ -0,0 +1,28 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+   <PropertyGroup>
+    <Authors>Microsoft.Marketplace</Authors>
+  	<Company>Microsoft</Company>
+    <Product>SaaS.SDK</Product>	  
+    <TargetFramework>net6.0.300</TargetFramework>
```
Comment (olsoro0) [inline src/SaaS.SDK.MeteredTriggerJob/SaaS.SDK.MeteredTriggerJob.csproj]: This produces a non-blocking error during the accelerator installation:
"error NETSDK1045: The current .NET SDK does not support targeting .NET 6.0.300.  Either target .NET 6.0 or lower,
 or use a version of the .NET SDK that supports .NET 6.0.300."

File: src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                    schedulerService.SaveSchedulerDetail(scheduler);
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+            }
+        }
+
+        public static void PrintScheduler(SchedulerManagerViewModel item, DateTime? nextRun)
```
[the remark this answers is not in the document]
Comment (msalemcode) [reply src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs:155]: Please check latest version it will compile now

**Review by tiffwchen — APPROVED** [review]

File: src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml [hunk 9]
```diff
@@ -21,8 +21,38 @@
...
+              date.min=output
+              $('#FirstRunDate').attr('min', output);
+              $('#FirstRunDate').val(output);
+              $('#FirstRunDate').attr('value',output);
+
+          });
+    }
+    );
+
+    function validateDate()
```
Comment (code4clouds) [inline src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml:46]: Is this validation also done on the server side/function?

File: .gitignore [hunk 10]
```diff
@@ -362,5 +362,8 @@ src/.idea/.idea.Microsoft.Marketplace.SaaS.SDK/.idea/projectSettingsUpdater.xml
 src/.idea/.idea.Microsoft.Marketplace.SaaS.SDK/.idea/vcs.xml
 
 # ignoring developer settings file
+appsettings.Development.json
+src/SaaS.SDK.PublisherSolution/appsettings.Development.json
+src/SaaS.SDK.MeteredTriggerJob/appsettings.Development.json
 src/SaaS.SDK.CustomerProvisioning/appsettings.Development.json
-src/SaaS.SDK.PublisherSolution/appsettings.Development.json
\ No newline at end of file
+src/SaaS.SDK.PublisherSolution/appsettings.Development.json
```
Comment (code4clouds) [inline .gitignore:369]: How this line different from 366?

File: deployment/Database/AMP-DB-5.1.sql [hunk 11]
```diff
@@ -407,6 +407,44 @@ CREATE TABLE [dbo].[Users](
...
+
+SET QUOTED_IDENTIFIER ON
+GO
+CREATE TABLE [dbo].[MeteredPlanSchedulerManagement] (
+    [Id]             INT          IDENTITY (1, 1) NOT NULL,
+    [SchedulerName]  VARCHAR (50) NOT NULL,
+    [SubscriptionId] INT          NOT NULL,
+    [PlanId]         INT          NOT NULL,
+    [DimensionId]    INT          NOT NULL,
+    [Quantity]       FLOAT (53)   NOT NULL,
```
Comment (code4clouds) [inline deployment/Database/AMP-DB-5.1.sql:423]: I thought we can only do uint in quantites...

File: deployment/Templates/Deploy.ps1 [hunk 12]
```diff
@@ -226,11 +227,32 @@ $URLToBacpacFromStorage = (Get-AzStorageBlob -blob $BlobName -Container $Contain
 Write-host "📜  Uploaded the bacpac file to $URLToBacpacFromStorage"    
 
 
+if (!($MeteredSchedulerSupport))
```
Comment (code4clouds) [inline deployment/Templates/Deploy.ps1:230]: Add a note to mentioned default behavior is to deploy meter scheduler.

File: deployment/Templates/Deploy.ps1 [hunk 13]
```diff
@@ -226,11 +227,32 @@ $URLToBacpacFromStorage = (Get-AzStorageBlob -blob $BlobName -Container $Contain
...
+
 Write-host "☁  Prepare publish files for the web application"
 
-Write-host "☁  Preparing the publish files for PublisherPortal"  
-dotnet publish ..\..\src\SaaS.SDK.PublisherSolution\SaaS.SDK.PublisherSolution.csproj -c debug -o ..\..\Publish\PublisherPortal
-Compress-Archive -Path ..\..\Publish\PublisherPortal\* -DestinationPath ..\..\Publish\PublisherPortal.zip -Force
+if ($MeteredSchedulerSupport -ne "NO")
+{ 
+    Write-host "☁  Preparing the publish files for PublisherPortal webjob"  
+    dotnet publish ..\..\src\SaaS.SDK.PublisherSolution\SaaS.SDK.PublisherSolution.csproj -c debug -o ..\..\Publish\PublisherPortal
```
Comment (code4clouds) [inline deployment/Templates/Deploy.ps1:240]: Line 240 and 252 seems to be the same, bubble this up outside of the if statement to simplify maintenance.

File: src/SaaS.SDK.PublisherSolution/Startup.cs [hunk 14]
```diff
@@ -122,6 +125,13 @@ public void ConfigureServices(IServiceCollection services)
 
             InitializeRepositoryServices(services);
 
+            services.AddDistributedMemoryCache();
```
Comment (code4clouds) [inline src/SaaS.SDK.PublisherSolution/Startup.cs:128]: Why are we adding session?

File: src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+using Microsoft.Marketplace.SaaS.SDK.Services.Models;
+using Microsoft.Marketplace.SaasKit.Client.DataAccess.Entities;
+using System.Text.Json;
+using Microsoft.Marketplace.SaaS.SDK.Services.Exceptions;
+using Microsoft.Marketplace.SaaS.SDK.Services.Contracts;
+using Microsoft.Marketplace.SaaS.SDK.Services.Services;
+using Microsoft.Marketplace.SaasKit.Client.DataAccess.Contracts;
+using Microsoft.Marketplace.SaaS.SDK.Services.Configurations;
+
+namespace MeteredTriggerHelper
```
Comment (code4clouds) [inline src/SaaS.SDK.MeteredTriggerJob/MeteredTriggerHelper.cs:14]: This whole file has no comments

**Review by code4clouds — CHANGES_REQUESTED** [review]

File: src/SaaS.SDK.PublisherSolution/Controllers/SchedulerController.cs [hunk 16]
```diff
@@ -254,7 +254,7 @@ public IActionResult SchedulerLogDetail(string id)
                 schedulerUsageViewModel.SelectedSchedulerFrequency = SchedulerItem.Frequency;
                 schedulerUsageViewModel.Quantity = (decimal)SchedulerItem.Quantity;
```
Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Controllers/SchedulerController.cs:255]: Can you change Quantity to be an integer (not double) in the viewmodel and then remove this cast?

File: src/SaaS.SDK.PublisherSolution/Controllers/SchedulerController.cs [hunk 17]
```diff
@@ -253,7 +254,7 @@ public IActionResult SchedulerLogDetail(string id)
                 schedulerUsageViewModel.SelectedDimension = SchedulerItem.Dimension;
                 schedulerUsageViewModel.SelectedSchedulerFrequency = SchedulerItem.Frequency;
                 schedulerUsageViewModel.Quantity = (decimal)SchedulerItem.Quantity;
-                schedulerUsageViewModel.FirstRunDate = SchedulerItem.StartDate;
+                schedulerUsageViewModel.FirstRunDate = SchedulerItem.StartDate.ToLocalTime();
```
Comment (olsoro0) [inline src/SaaS.SDK.PublisherSolution/Controllers/SchedulerController.cs:257]: Why did you change this to be local time? Should it be UTC?

File: src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml (hunk 9 above)
Comment (msalemcode) [reply src/SaaS.SDK.PublisherSolution/Views/Scheduler/NewScheduler.cshtml]: the date validation is on the client side beside there are validation built in the HTML 5 input itself

**Review by code4clouds — APPROVED** [review]

File: docs/Installation-Instructions.md [hunk 18]
```diff
@@ -1,15 +1,11 @@
...
   
    1. Copy the following section to an editor and update it to match your company preference. Replace `SOME-UNIQUE-STRING` with your Team name or some other random string.
 
 ``` powershell
-wget https://dotnet.microsoft.com/download/dotnet/scripts/v1/dotnet-install.sh; `
-chmod +x dotnet-install.sh; `
-./dotnet-install.sh; `
-$ENV:PATH="$HOME/.dotnet:$ENV:PATH"; `
-git clone https://github.com/Azure/Commercial-Marketplace-SaaS-Accelerator.git -b 5.0.0 --depth 1; `
+git clone https://github.com/Azure/Commercial-Marketplace-SaaS-Accelerator.git -b 5.1.0 --depth 1; `
```
[the remark this answers is not in the document]
Comment (code4clouds) [reply docs/Installation-Instructions.md:8]: I will approve and change directly.

**Pull request merged**